### ISSUE TITLE
"Unable to create a signed JWT from the given configuration" on Symfony test project deployed at Heroku's

### ISSUE THREAD

**Issue by juliolopeztorres** [lexik/LexikJWTAuthenticationBundle#839]
Hi there!

First of all, thanks for your work and for this package.

I am trying to test a tiny Symfony 5.2 backend on Heroku's free plan but I face the _usual_ (I mean, there are some closed issues talking about this) problem when I hit `login`. In dev environment (local machine) everything work smoothly. I double-check path to `.pem`, files permissions on Heroku's container, the passphrase, regenerate the `.pem` manually as you recommended in older version of the package as well as trying to generate it with the new command, everything :(

I always got a 500 telling me that "unable to created signed JWT" at `LcobucciJWTEncoder` class, then calling `->isSigned()` method. I also tried older version the your package such as 2.9.5 but still nearly the same result (""An error occurred while trying to encode the JWT token. Please verify your configuration (private key/passphrase)" at /app/vendor/lexik/jwt-authentication-bundle/Encoder/LcobucciJWTEncoder.php line 37"). 

This is driving me crazy. 

To sum up: 

- Problem when answering the `/login` path
- Tried package version 2.11.1
- Tried package version 2.11.0
- Tried package version 2.9.5
- Symfony 5.2 and PHP 7.4

**Comment by chalasr** [comment]
You may need to set the corresponding environment variables in your heroku env, you can do it via the Heroku UI (see https://github.com/lexik/LexikJWTAuthenticationBundle/issues/426#issuecomment-360084850).

**Comment by juliolopeztorres** [comment]
Thank you for your response.

I did write the `env` variables both through Heroku CLI and on Heroku GUI, double-checking its content. I think the problem is not that the `.env` does not exist or the path is wrong, based on the error I got using the last release of the package. Besides, I ended up hardcoding the routes and the password on `parameters` just in case it was an environment-parameters issue. No luck. 

I also play with the routes and I checked that the pathing is correct. The error seems arise in the `isSigned()` method in `v2.11.1`.

**Comment by chalasr** [comment]
Do you mean it work as expected with versions prior to v2.11.1?

**Comment by juliolopeztorres** [comment]
No. I did several tries based on the doc from the sandbox repo and from other tickets like (https://github.com/lexik/LexikJWTAuthenticationBundle/issues/532). 

I have tried again with `v2.11.1` on the bundle, generating the keypair with the `lexik:jwt:generate-keypair` as well as by CLI using the info (https://github.com/lexik/LexikJWTAuthenticationBundle/blob/v2.9.0/Resources/doc/index.md) and `openssl`.

Still no luck. I always get:

`Uncaught PHP Exception Lexik\Bundle\JWTAuthenticationBundle\Exception\JWTEncodeFailureException: "Unable to create a signed JWT from the given configuration." at /app/vendor/lexik/jwt-authentication-bundle/Encoder/LcobucciJWTEncoder.php line 38`.

That error implied that:

```        try {
            $token = $this->getSignedToken($jws);
        } catch (\InvalidArgumentException $e) {
        }

        return new CreatedJWS((string) $token, null === $e);```

throws an exception as the $e is not null when `CreatedJWS` is called on `LcobucciJWSProvider` (`v4.1.0`), lines 127-132.

If I run `lexik:jwt:check-config`, it said that `the configuration seems correct` and with `lexik:jwt:generate-token` I can generate tokens but no luck when hitting `/login` :(

**Comment by juliolopeztorres** [comment]
Hi there again. 

After several testing locally and messing around with configuration both in the `.yaml` and in my `Heroku` project dashboard, I ended up with something like #634. 

My main problem was to think that I could connect to `Heroku` through `heroku run bash` and execute there the commands `lexik:jwt:generate-token` or create the `.pem`, among others, manually. 

I still don't understand what is happening behind the scenes in the deployment process but I have emptied almost all my environment variables (preserving `database connection`, `APP_URL` and `APP_ENV`)  on `Heroku`'s project dashboard. 

My solution:

The symfony package `dotenv` is required by flex in the current implementation of `symfony/skeleton` (for `symfony` 5.2) which I took as start point. When I push to `Heroku`, the Symfony recipes create a copy of the `.env.dist`, generating another file `.env`. A picture of the interested part from mine is:

```
###> lexik/jwt-authentication-bundle ###
JWT_SECRET_KEY=%kernel.project_dir%/config/jwt/private.pem
JWT_PUBLIC_KEY=%kernel.project_dir%/config/jwt/public.pem
JWT_PASSPHRASE=
###< lexik/jwt-authentication-bundle ###
```

The configuration for the bundle ended up like:

```
lexik_jwt_authentication:
    secret_key: '%env(resolve:JWT_SECRET_KEY)%'
    public_key: '%env(resolve:JWT_PUBLIC_KEY)%'
    pass_phrase: '%env(JWT_PASSPHRASE)%'
```

As you can see, nearly the same than the content that the own recipe generates from your package. 

Finally, the file trick is to execute the creation of the `config/jwt` folder and the `private.pem`/`public.pem` inside the `post-install` hook, as #634 mention. In my case:

```
        "post-install-cmd": [
            "@auto-scripts",
            "mkdir -p config/jwt",
            "openssl genrsa -out config/jwt/private.pem 4096",
            "openssl rsa -pubout -in config/jwt/private.pem -out config/jwt/public.pem",
            "chmod -R 775 config/"
        ],
```

I hope someone out there might find this useful and sorry for the inconvenience.

**Comment by Volmarg** [comment]
So this `php bin/console lexik:jwt:generate-keypair` didn't worked.
Yet the mentioned by @juliolopeztorres 

```
            "openssl genrsa -out config/jwt/private.pem 4096",
            "openssl rsa -pubout -in config/jwt/private.pem -out config/jwt/public.pem",
```

works

**Comment by tatahappyness** [comment]
Help me please to solve this problem in symfony 5.2 :

Unable to create a signed JWT from the given configuration. (500 Internal Server Error)

**Comment by tomgif** [comment]
> Help me please to solve this problem in symfony 5.2 :
> 
> Unable to create a signed JWT from the given configuration. (500 Internal Server Error)

i got the same issue with manually created keys.
I solved it by using `php bin/console lexik:jwt:generate-keypair --overwrite`

**Comment by JCSama** [comment]
I had the same problem so according the Heroku doc you can add a compile step within composer.json file: 
```
"scripts": {
       ...
        "compile": [
            "@php bin/console lexik:jwt:generate-keypair --overwrite --quiet --env prod"
        ]
    }
```

**Comment by sunviwo** [comment]
> "compile": [
>             "@php bin/console lexik:jwt:generate-keypair --overwrite --quiet --env prod"
>         ]

This solution worked for me. Thanks

**Comment by Jehovanie** [comment]
Thank's... 🤗
